### PR TITLE
feat: source post moderation cdc - post creation

### PR DESCRIPTION
**Pull request by sshanzel**
Changes:
- CDC that manages the creation of posts.
- The RFC may have specified on post delete, we need to delete the relevant requests, but we won't need it anymore since we cascade on delete on their relation.
- Refactored some functions to make it reusable to creating posts.

### CODE REVIEW

**Comment by sshanzel** [comment]
So glad the tests have passed 😌 

Time to write a ton of tests now for the CDC.

File: src/common/post.ts
Comment (rebelchris) [inline src/common/post.ts:362]: This is just for failing if doesn't exist? (maybe add explicit id select to narrow scope)

File: src/common/post.ts
Comment (rebelchris) [inline src/common/post.ts:371]: Don't we do stuff with markdown or tags here? (can't recall so top of mind)

File: src/common/post.ts
Comment (rebelchris) [inline src/common/post.ts:375]: Same here if cloudinary upload etc

File: src/common/post.ts
Comment (rebelchris) [inline src/common/post.ts:420]: Is this was we do now?
I think @ilasw mentioned it in another thread as well.

File: src/entity/posts/utils.ts
Comment (rebelchris) [inline src/entity/posts/utils.ts:308]: Why do we need it returned now?

File: src/entity/posts/utils.ts
Comment (rebelchris) [inline src/entity/posts/utils.ts:399]: @omBratteng was there specific case you checked === true?

**Review by rebelchris — COMMENTED** [review]
Overall minor comments so looking good.

I see we have a ton of `create post` kind of functions is there no way we can somehow combine them into seperated utility functions?
I find it very hard to follow why there's multiple things that all look similar.

CC: @capJavert for your opinion as well?

File: src/common/post.ts
Comment (capJavert) [inline src/common/post.ts:465]: Let's use object style syntax as we use in other workers, it will give us better visibility

```suggestion
  logger.error({ moderated }, 'unable to process moderated post');
```

**Review by capJavert — APPROVED** [review]
Looks good, some sanity checks, but good to merge after!

File: src/common/post.ts
Comment (capJavert) [inline src/common/post.ts:383]: @rebelchris I agree with duplicated functions, but I think it just comes to the fact we often add utility functions in different files without really looking if they are used or if there is something similar. 

Would say this is an example, its used in a single place and for sure if I need to check existing post I'll just write it myself again.

**Comment by sshanzel** [comment]
> I see we have a ton of create post kind of functions is there no way we can somehow combine them into seperated utility functions?

Totally agree with that, although I tried to make them generic as much as I can, but we have three different processes for 3 different kind of submissions. For Freeform, we just create everything the user adds, for share, we have two kinds. One is sharing already existing post, the other accepts external link. All 3 of them have different requirements, for example, Share does not need `content`, then on external link we have a `title` but going to be used on the article post generated then generate another share post with the `content`.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:222]: The function returns seem to be consistent to be a boolean type. Feel free to let me know if I have to revert.

File: src/entity/posts/utils.ts
Comment (sshanzel) [reply src/entity/posts/utils.ts:308]: We now need to return it. This makes us reuse the same function. Previous usages should not matter whether this returns anything or not.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:420]: This is for the actual post in the `post` table, for the moderation table, we generate it in DB level. If that is what you mean as I don't have reference to the thread.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:362]: I'm not sure I understand. We have `id: moderated.postId` already right?

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:371]: Yes, and since from the post submission we generate it already. Hence, we just pass it here.

Comment (rebelchris) [reply src/common/post.ts:371]: Ah ok, guess just very hard to follow from review which part actually does something 😢 
One day we need to clean this up a bit I feel.

File: src/common/post.ts
Comment (rebelchris) [reply src/common/post.ts:420]: Ah maybe yes, this one we did now with shortId?

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:362]: Oh, didn't know you actually pull everything if you don't specify. Yeah, sure let's do that.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:371]: Yeah, totally agree. Feels like we need some sort of process flow that will handle everything when it comes to creating post.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:420]: I just saw from the other implementation where we create post we use this id. Wdyt, should we use uuid?

File: src/entity/posts/utils.ts
Comment (sshanzel) [reply src/entity/posts/utils.ts:308]: Because the function is now used on two parts, and the other one needs to utilize some prop from the created post. Specifically the id.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:211]: Yeah, just wanted to avoid huge chunks of indent, but we can do that too.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:465]: Cool. Didn't wanna flood or logs with might be unnecessary props which can just be pulled.

File: src/common/post.ts
Comment (rebelchris) [reply src/common/post.ts:383]: Yeah probably good idea, it's not blocking for now, we can log it for web guild meeting :) 
It's good to think about these systems to avoid re-building stuff.

File: src/common/post.ts
Comment (rebelchris) [reply src/common/post.ts:420]: I personally don't mind more up to @capJavert I think

File: src/common/post.ts
Comment (rebelchris) [reply src/common/post.ts:371]: Let's log it for backlog, don't want to keep this project blocked by it.

Comment (rebelchris) [reply src/common/post.ts:371]: @sshanzel 

> Totally agree with that, although I tried to make them generic as much as I can, but we have three different processes for 3 different kind of submissions. For Freeform, we just create everything the user adds, for share, we have two kinds. One is sharing already existing post, the other accepts external link. All 3 of them have different requirements, for example, Share does not need content, then on external link we have a title but going to be used on the article post generated then generate another share post with the content.

This is a very good point, ideally this is just wrapped in 3 functions then:
- `createFreeformPost`
- `createShareExistingPost`
- `createShareExternalLinkPost`

And those rule all the things, but it's more tech debt clean up as this point.

File: src/common/post.ts
Comment (capJavert) [reply src/common/post.ts:465]: This should basically never happen, so if when deploy it starts to happen we need to adjust the worker so it handles the cases which end up on this

**Review by ilasw — APPROVED** [review]
Nice, looks really good :)

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:465]: Makes sense, yeah. I'll update quickly.

File: src/common/post.ts
Comment (sshanzel) [reply src/common/post.ts:373]: Yeah, that would be best. We actually apply that now here, hopefully, we can apply it on the legacy code too. 

Ticket created: https://dailydotdev.atlassian.net/jira/software/projects/WT/boards/11/backlog?selectedIssue=WT-2245

**Review by rebelchris — APPROVED** [review]

**Pull request merged**
